### PR TITLE
Extensions.Hosting fixes

### PR DESCRIPTION
**Pull request by cijothomas**
Replaces previous PRs: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2006,  https://github.com/open-telemetry/opentelemetry-dotnet/pull/1998, https://github.com/open-telemetry/opentelemetry-dotnet/pull/2004

1. Only allow AddOpenTelemetryTracing() to be called once. Attempting to call more than once will throw. In other words, Ext.hosting only allows a single TracerProvider to be registered, and this is the common usage pattern.
   If user has a need to have multiple TracerProvider instance, they can still create it and add to DI themselves.
This resolves https://github.com/open-telemetry/opentelemetry-dotnet/issues/1215
2. The AddOpenTelemetryTracing(), now picks up Sampler, Processor from DI, without requiring users to call the `SetSampler<T>`, `AddProcessor<T>` extension methods from the Ext.Hosting, and hence these methods are removed.

```
               services.AddOpenTelemetryTracing();
               // The following Sampler is picked up.
               services.AddSingleton<Sampler>(new ParentBasedSampler(new AlwaysOffSampler()));
               // The following Processor is picked up.
               services.AddSingleton<BaseProcessor<Activity>>(new TestProcessor());
```

This will address https://github.com/open-telemetry/opentelemetry-dotnet/issues/831 partially. The below TODO will address it completely.

TODO (in a future PR) - introduce a base class for instrumentations to enable instrumentations to be also added similar to above.
TODO : More tests, and an updated documentation. Want to see if this change is acceptable, before spending more time here.
TODO: Changelog.md updates (to be done after getting a +1 to this approach)

### CODE REVIEW

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs [hunk 1]
```diff
@@ -50,6 +51,11 @@ public static IServiceCollection AddOpenTelemetryTracing(this IServiceCollection
                 throw new ArgumentNullException(nameof(configure));
             }
 
+            if (services.Any(service => service.ImplementationType == typeof(TelemetryHostedService)))
+            {
+                throw new InvalidOperationException("AddOpenTelemetryTracing must be only called once.");
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:56]: I feel like a `NotSupportedException` might be slightly better here, something like:

```csharp
throw new NotSupportedException("Registering multiple TracerProviders via AddOpenTelemetryTracing is not supported.");
```

File: src/OpenTelemetry.Extensions.Hosting/README.md [hunk 2]
```diff
@@ -29,20 +29,23 @@ services.Configure<ZipkinExporterOptions>(this.Configuration.GetSection("Zipkin"
...
 services.AddOpenTelemetryTracing((builder) => builder
     .AddAspNetCoreInstrumentation()
-    .AddHttpClientInstrumentation()
-    .AddProcessor<MyProcessor>());
+    .AddHttpClientInstrumentation());
 ```
 
-Similar methods exist for registering instrumentation (`AddInstrumentation<T>`)
-and setting a sampler (`SetSampler<T>`).
+Similarly, sampler can also be set of the `IServiceCollection`.
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/README.md:44]: typo: Similarly, sampler can also be set **on** the `IServiceCollection`.

File: src/OpenTelemetry.Extensions.Hosting/README.md [hunk 3]
```diff
@@ -29,20 +29,23 @@ services.Configure<ZipkinExporterOptions>(this.Configuration.GetSection("Zipkin"
...
 
-The following example registers a processor of the type "MyProcessor" which has
-been registered as a singleton with the `IServiceCollection`:
+The following example adds a processor instance
+as a singleton to the `IServiceCollection` which will
+be picked up by the TracerProvider.
 
 ```csharp
-services.AddSingleton<MyProcessor>();
+services.AddSingleton<BaseProcessor<Activity>>(new MyProcessor());
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/README.md:37]: Maybe add a second one to demo you can register multiple?

File: src/OpenTelemetry.Extensions.Hosting/README.md [hunk 4]
```diff
@@ -29,20 +29,23 @@ services.Configure<ZipkinExporterOptions>(this.Configuration.GetSection("Zipkin"
...
+    .AddHttpClientInstrumentation());
 ```
 
-Similar methods exist for registering instrumentation (`AddInstrumentation<T>`)
-and setting a sampler (`SetSampler<T>`).
+Similarly, sampler can also be set of the `IServiceCollection`.
+
+```csharp
+services.AddSingleton<Sampler>(new TestSampler());
+```
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/README.md:48]: FYI below this in the "Building Extension Methods" section there are a couple more calls to AddProcessor/SetSampler.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs [hunk 5]
```diff
@@ -50,6 +51,11 @@ public static IServiceCollection AddOpenTelemetryTracing(this IServiceCollection
                 throw new ArgumentNullException(nameof(configure));
             }
 
+            if (services.Any(service => service.ImplementationType == typeof(TelemetryHostedService)))
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:54]: I was thinking about this a bit more. We have to loop through all the services in order to make this determination. That is going to slow down startup which will be especially impactful to things like functions & lambdas. We might want to just introduce a static we can check using Interlocked.CompareExchange?

File: src/OpenTelemetry.Extensions.Hosting/README.md [hunk 6]
```diff
@@ -79,12 +83,10 @@ public static class MyLibraryExtensions
                 "MyFeature requires a hosting TracerProviderBuilder instance."))
             .AddHostedService<MyHostedService>()
             .AddSingleton<MyService>()
-            .AddSingleton<MyProcessor>()
-            .AddSingleton<MySampler>();
+            .AddSingleton<BaseProcessor<Activity>>(new MyProcessor());
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/README.md:86]: This signature isn't very nice, maybe we should add an extension in hosting something like...

```csharp
        public static IServiceCollection AddActivityProcessor<T>(this IServiceCollection services, ServiceLifetime serviceLifetime = ServiceLifetime.Singleton)
            where T : BaseProcessor<Activity>
        {
            services.Add(ServiceDescriptor.Describe(typeof(BaseProcessor<Activity>), typeof(T), serviceLifetime));
            return services;
        }
```

Comment (cijothomas) [reply src/OpenTelemetry.Extensions.Hosting/README.md:86]: yes, agree with that.

File: src/OpenTelemetry.Extensions.Hosting/README.md [hunk 7]
```diff
@@ -29,20 +29,24 @@ services.Configure<ZipkinExporterOptions>(this.Configuration.GetSection("Zipkin"
...
-    .AddProcessor<MyProcessor>());
+    .AddHttpClientInstrumentation());
 ```
 
-Similar methods exist for registering instrumentation (`AddInstrumentation<T>`)
-and setting a sampler (`SetSampler<T>`).
+Similarly, sampler can also be set on the `IServiceCollection`.
+
+```csharp
+services.AddSingleton<Sampler>(new TestSampler());
```
Comment (reyang) [inline src/OpenTelemetry.Extensions.Hosting/README.md:48]: If there are multiple invocations, who would be responsible for disposing the unused one? e.g.:
```csharp
class FooSampler: IDisposable, Sampler
{
}

services.AddSingleton<Sampler>(new FooSampler("A")); // do we leak this?
services.AddSingleton<Sampler>(new FooSampler("B")); // do we always pick the `last one`? or we want to throw exception if there is more than one?
```

https://github.com/open-telemetry/opentelemetry-dotnet/blob/146dd73ed28ddc94b8335306eb62dff5337d783c/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L310

**Pull request closed without merging**